File: src/main/java/fr/cph/crypto/controller/TickerController.java
package fr.cph.crypto.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import fr.cph.crypto.domain.Currency;
import fr.cph.crypto.domain.Ticker;
import fr.cph.crypto.repository.TickerRepository;

@RequestMapping(value = "/ticker")
@RestController
public class TickerController {

	private final TickerRepository repository;

	@Autowired
	public TickerController(final TickerRepository repository) {
		this.repository = repository;
	}

	@RequestMapping
	public List<Ticker> getAll() {
		return StreamSupport.stream(repository.findAll().spliterator(), false).collect(Collectors.toList());
	}

	@RequestMapping(value = "/{baseCurrency}/{quoteCurrency}")
	public Ticker getTicker(
			@PathVariable("baseCurrency") final Currency baseCurrency,
			@PathVariable("quoteCurrency") final Currency quoteCurrency) {
		return repository.findOne(baseCurrency.getCode() + "-" + quoteCurrency.getCode());
	}
}

File: src/main/kotlin/fr/cph/crypto/controller/TickerController.kt
package fr.cph.crypto.controller

import fr.cph.crypto.domain.Currency
import fr.cph.crypto.domain.Ticker
import fr.cph.crypto.service.TickerService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController

@RequestMapping(value = ["/api/ticker"])
@RestController
class TickerController @Autowired
constructor(private val service: TickerService) {

    @PreAuthorize("hasAuthority('ADMIN') or authentication.details.decodedDetails['id'] == null")
    @RequestMapping(method = [RequestMethod.GET])
    fun getAllTickers(): List<Ticker> {
        return service.findAll()
    }

    @PreAuthorize("hasAuthority('ADMIN') or authentication.details.decodedDetails['id'] == null")
    @RequestMapping(value = ["/{currency1}/{currency2}"], method = [RequestMethod.GET])
    fun getTicker(@PathVariable("currency1") currency1: Currency,
                  @PathVariable("currency2") currency2: Currency): Ticker {
        return service.findOne(currency1.code + "-" + currency2.code)
    }
}
